source/OrderedTypes/PredicativeLFP.lagda

```diff
@@ -0,0 +1,1704 @@
+a least fixed point.
+
+\begin{code}
+
+module _ {𝓤 𝓦 𝓥 : Universe} (L : Sup-Lattice 𝓤 𝓦 𝓥) where
+
+ has-least-fixed-point : (f : ⟨ L ⟩ → ⟨ L ⟩) → 𝓤 ⊔ 𝓦  ̇
+ has-least-fixed-point f =
+  Σ p ꞉ ⟨ L ⟩ , (f p ＝ p) × ((a : ⟨ L ⟩) → (f a ＝ a) → (p ≤⟨ L ⟩ a) holds)
+
+ has-least-fixed-point-is-prop : (f : ⟨ L ⟩ → ⟨ L ⟩)
+                               → is-prop (has-least-fixed-point f) 
+ has-least-fixed-point-is-prop f (p₁ , fp₁ , l₁) (p₂ , fp₂ , l₂) =
+  to-subtype-＝ (λ x → ×-is-prop
+                       (sethood-of L)
+                       (Π-is-prop fe (λ y → Π-is-prop fe
+                        (λ _ → holds-is-prop (x ≤⟨ L ⟩ y)))))
+                (antisymmetry-of L (l₁ p₂ fp₂) (l₂ p₁ fp₁))
+
+\end{code}
+
+We construct the least closed subset of an inductive definition as a QIT family.
+Since QITs (and more generally HITs) are not native in Agda we will instead
+assume the existence of such a type as well as its induction principle.
+Technically speaking we are going to use record types to package the contents
+of this QIT family.
+
+Notice all constructors are 'strictly positive' and the path constructors are
+'natural' (see Chapter 6 Section 13 of the HoTT book
+https://homotopytypetheory.org/book/).
+
+Notice that the QIT family has two constructors which represent the closure
+conditions we wish to impose on subsets. The c-closure condition says:
+for any subset contained in the least closed subset, elements in the downset of
+its join are in the least closed subset as well. That is, Y is c-closed if
+  for any U ⊆ Y we have ↓ᴮ (⋁ U) ⊆ Y.
+The ϕ-cl constructor says:
+if for any a : L and b : B with (b , a) ∈ ϕ and ↓ᴮ a 'contained' in the least
+closed subset then b is in the least closed subset. That is, Y is ϕ-closed if
+  for any a : L and b : B we have ↓ᴮ a ⊆ Y ⇒ b ∈ Y.
+
+It is worth noting that we don't encode the downsets as subsets in type
+theory (rather they are total spaces) so for that reason we won't encode the
+closure conditions exactly as above (maybe add some notation to allow for
+more familiar form).
+
+We also derive the initiality of the least closed subset from the postulated
+induction principle. Initiality is closely related to the 'least' part of
+our least fixed point theorem.
+
+\begin{code}
+
+module _
+        {𝓤 𝓦 𝓥 : Universe}
+        {B : 𝓥  ̇}
+        (L : Sup-Lattice 𝓤 𝓦 𝓥)
+        (β : B → ⟨ L ⟩)
+        (h : is-small-basis L β)
+       where
+
+ private
+  _≤_ = order-of L
+  ⋁_ = join-of L
+
+ open Joins _≤_
+ open is-small-basis h
+
+\end{code}
+
+The following record type should be interpreted as supplying the assumption
+that the QIT family exists with the apropriate induction principle.
+
+\begin{code}
+
+ record inductively-generated-subset-exists (ϕ : 𝓟 {𝓤 ⊔ 𝓥} (B × ⟨ L ⟩)): 𝓤ω
+  where
+  constructor
+   inductively-generated-subset
+
+  field
+   Ind : B → 𝓤 ⊔ 𝓥 ⁺  ̇
+   Ind-trunc : (b : B) → is-prop (Ind b)
+   c-closed : (U : 𝓟 {𝓥} B)
```

Comment:
Might it be possible to define `c-closure` and `ϕ-closure` so that you can refer to these types both here and in `Ind-induction`? I'm thinking of something like `c-closed : c-closure Ind`.

Reply:
This (or a minor variation) should ideally also clean up the next code block. The types there are rather long.

Reply:
Yes this is a good idea. Although the induction principles contains something like dependent closure properties. I guess I could make the type c-closure mirror this dependent structure but it feels a bit like a misnomer at that point.

Reply:
Alternatively I could make a type called dep-c-closure in addition to normal c-closure. Any thoughts on this? @tomdjong